src/deepsparse/utils/timer.py

```diff
@@ -0,0 +1,202 @@
+
+
+class StagedTimer:
+    def __init__(self):
+        self._staged_start_times = {}
+        self._staged_stop_times = {}
+
+    def __repr__(self):
+        return f"InferenceTimer({self.times})"
+
+    @property
+    def stages(self) -> List[str]:
+        return list(self._staged_start_times.keys())
+
+    @property
+    def times(self) -> Dict[str, float]:
+        return {stage: self.stage_average_time(stage) for stage in self.stages}
+
+    @property
+    def all_times(self) -> Dict[str, List[float]]:
+        return {stage: self.stage_times(stage) for stage in self.stages}
+
+    def clear(self):
+        self._staged_start_times.clear()
+        self._staged_stop_times.clear()
+
+    def has_stage(self, stage: str) -> bool:
+        return stage in self.stages
+
+    def start(self, stage: str):
+        if stage not in self._staged_start_times:
+            self._staged_start_times[stage] = []
+            self._staged_stop_times[stage] = []
+
+        if len(self._staged_start_times[stage]) != len(self._staged_stop_times[stage]):
+            raise ValueError(
+                f"Attempting to start {stage} before a previous has been stopped:"
+                f" start times len({self._staged_start_times[stage]});"
+                f" stop times len({self._staged_stop_times[stage]})"
+            )
+
+        self._staged_start_times[stage].append(time.perf_counter())
+
+    def stop(self, stage: str):
+        if stage not in self._staged_start_times:
+            raise ValueError(
+                "Attempting to stop a stage that has not been started: " f"{stage}"
+            )
+
+        if (
+            len(self._staged_start_times[stage])
+            != len(self._staged_stop_times[stage]) + 1
+        ):
+            raise ValueError(
+                f"Attempting to stop {stage} before a previous has been started:"
+                f" start times len({self._staged_start_times[stage]});"
+                f" stop times len({self._staged_stop_times[stage]})"
+            )
+
+        self._staged_stop_times[stage].append(time.perf_counter())
+
+    def stage_times(self, stage: str) -> List[float]:
+        if stage not in self._staged_start_times:
+            raise ValueError(
+                "Attempting to get time deltas for a stage that has not been started: "
+                f"{stage}"
+            )
+
+        if len(self._staged_start_times[stage]) != len(self._staged_stop_times[stage]):
+            raise ValueError(
+                "Attempting to get time deltas for a stage that has not been stopped: "
+                f"{stage}"
+            )
+
+        return [
+            self._staged_stop_times[stage][i] - self._staged_start_times[stage][i]
+            for i in range(len(self._staged_start_times[stage]))
+        ]
+
+    def stage_average_time(self, stage: str) -> float:
+        times = self.stage_times(stage)
+
+        return sum(times) / len(times)
+
+
+class InferenceTimer:
```

Comment:
Missing docstrings. It was not initially obvious to me what e.g. the `multi` argument means.

Reply:
docstrings added